<a href="https://colab.research.google.com/github/nsriniva/DS-Unit-1-Sprint-2-Statistics/blob/master/DS_Unit_1_Sprint_Challenge_2_Statistics.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Science Unit 1 Sprint Challenge 2

## Exploring Data, Testing Hypotheses

In this sprint challenge you will look at a dataset of people being approved or rejected for credit.

https://archive.ics.uci.edu/ml/datasets/Credit+Approval

Data Set Information: This file concerns credit card applications. All attribute names and values have been changed to meaningless symbols to protect confidentiality of the data. This dataset is interesting because there is a good mix of attributes -- continuous, nominal with small numbers of values, and nominal with larger numbers of values. There are also a few missing values.

Attribute Information:
- A1: b, a.
- A2: continuous.
- A3: continuous.
- A4: u, y, l, t.
- A5: g, p, gg.
- A6: c, d, cc, i, j, k, m, r, q, w, x, e, aa, ff.
- A7: v, h, bb, j, n, z, dd, ff, o.
- A8: continuous.
- A9: t, f.
- A10: t, f.
- A11: continuous.
- A12: t, f.
- A13: g, p, s.
- A14: continuous.
- A15: continuous.
- A16: +,- (class attribute)

Yes, most of that doesn't mean anything. A16 is a variable that indicates whether or not a person's request for credit has been approved or denied. This is a good candidate for a y variable since we might want to use the other features to predict this one. The remaining variables have been obfuscated for privacy - a challenge you may have to deal with in your data science career.

Sprint challenges are evaluated based on satisfactory completion of each part. It is suggested you work through it in order, getting each aspect reasonably working, before trying to deeply explore, iterate, or refine any given step. Once you get to the end, if you want to go back and improve things, go for it!

## Part 1 - Load and validate the data

- Load the data as a `pandas` data frame.
- Validate that it has the appropriate number of observations (you can check the raw file, and also read the dataset description from UCI).
- UCI says there should be missing data - check, and if necessary change the data so pandas recognizes it as na
- Make sure that the loaded features are of the types described above (continuous values should be treated as float), and correct as necessary

This is review, but skills that you'll use at the start of any data exploration. Further, you may have to do some investigation to figure out which file to load from - that is part of the puzzle.

In [13]:
# TODO
from re import compile as rcompile
names_text = !curl https://archive.ics.uci.edu/ml/machine-learning-databases/credit-screening/crx.names

rex = rcompile('(A[0-9]+):')
# Create a generator
matches  = (rex.search(line) for line in names_text)
col_names = [ match.group(1) for match in matches if match ]

# We cheat a bit here - fine tuning the regex isn't worth the effort so we just 
# second set of attributes from the missing values section
col_names=col_names[:16]
print('\n', len(col_names), col_names)



 16 ['A1', 'A2', 'A3', 'A4', 'A5', 'A6', 'A7', 'A8', 'A9', 'A10', 'A11', 'A12', 'A13', 'A14', 'A15', 'A16']


In [14]:
import pandas as pd
import numpy as np

credit_url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/credit-screening/crx.data'
credit_df = pd.read_csv(credit_url,header=None, names=col_names)

In [15]:
display(credit_df.head())
credit_df.shape

A1     A2     A3 A4 A5 A6 A7    A8 A9 A10  A11 A12 A13    A14  A15 A16
0  b  30.83  0.000  u  g  w  v  1.25  t   t    1   f   g  00202    0   +
1  a  58.67  4.460  u  g  q  h  3.04  t   t    6   f   g  00043  560   +
2  a  24.50  0.500  u  g  q  h  1.50  t   f    0   f   g  00280  824   +
3  b  27.83  1.540  u  g  w  v  3.75  t   t    5   t   g  00100    3   +
4  b  20.17  5.625  u  g  w  v  1.71  t   f    0   f   s  00120    0   +

(690, 16)

The number of observations(690) from the imported dataset matches the number of
observations reported in the UCI dataset description.

In [22]:
# We know from the names file that the following attributes have missing 
# information.
# A1:  12
# A2:  12
# A4:   6
# A5:   6
# A6:   9
# A7:   9
# A14: 13

# Examining value_counts for each attribute shows that the missing values have
# '?'. 
# Replacing '?' with NaN

credit_df = credit_df.replace('?',np.NaN)

#Verify that all missing data has been replaced with NaN
credit_df.isnull().sum()

A1     12
A2     12
A3      0
A4      6
A5      6
A6      9
A7      9
A8      0
A9      0
A10     0
A11     0
A12     0
A13     0
A14    13
A15     0
A16     0
dtype: int64

In [24]:
credit_df.dtypes

A1      object
A2      object
A3     float64
A4      object
A5      object
A6      object
A7      object
A8     float64
A9      object
A10     object
A11      int64
A12     object
A13     object
A14     object
A15      int64
A16     object
dtype: object

In [36]:
# From the names file A2, A3, A8, A11, A14 and A15 are continuous(aka float) 
# valued.
# We need to convert A2(object->float), A11(int->float), A14(object->float)
# and A15(int->float)

floats = ['A2','A3','A8','A11','A14','A15']
floats_conv = ['A2','A11','A14','A15']
for attr in floats_conv:
  credit_df[attr] = credit_df[attr].apply(float)

credit_df.dtypes

A1      object
A2     float64
A3     float64
A4      object
A5      object
A6      object
A7      object
A8     float64
A9      object
A10     object
A11    float64
A12     object
A13     object
A14    float64
A15    float64
A16     object
dtype: object

In [37]:
credit_df

A1     A2      A3 A4 A5  A6  A7    A8 A9 A10  A11 A12 A13    A14    A15 A16
0    b  30.83   0.000  u  g   w   v  1.25  t   t  1.0   f   g  202.0    0.0   +
1    a  58.67   4.460  u  g   q   h  3.04  t   t  6.0   f   g   43.0  560.0   +
2    a  24.50   0.500  u  g   q   h  1.50  t   f  0.0   f   g  280.0  824.0   +
3    b  27.83   1.540  u  g   w   v  3.75  t   t  5.0   t   g  100.0    3.0   +
4    b  20.17   5.625  u  g   w   v  1.71  t   f  0.0   f   s  120.0    0.0   +
..  ..    ...     ... .. ..  ..  ..   ... ..  ..  ...  ..  ..    ...    ...  ..
685  b  21.08  10.085  y  p   e   h  1.25  f   f  0.0   f   g  260.0    0.0   -
686  a  22.67   0.750  u  g   c   v  2.00  f   t  2.0   t   g  200.0  394.0   -
687  a  25.25  13.500  y  p  ff  ff  2.00  f   t  1.0   t   g  200.0    1.0   -
688  b  17.92   0.205  u  g  aa   v  0.04  f   f  0.0   f   g  280.0  750.0   -
689  b  35.00   3.375  u  g   c   h  8.29  f   f  0.0   t   g    0.0    0.0   -

[690 rows x 16 columns]

## Part 2 - Exploring data, Testing hypotheses

The only thing we really know about this data is that A16 is the class label. Besides that, we have 6 continuous (float) features and 9 categorical features.

Explore the data: you can use whatever approach (tables, utility functions, visualizations) to get an impression of the distributions and relationships of the variables. In general, your goal is to understand how the features are different when grouped by the two class labels (`+` and `-`).

For the 6 continuous features, how are they different when split between the two class labels? Choose two features to run t-tests (again split by class label) - specifically, select one feature that is *extremely* different between the classes, and another feature that is notably less different (though perhaps still "statistically significantly" different). You may have to explore more than two features to do this.

For the categorical features, explore by creating "cross tabs" (aka [contingency tables](https://en.wikipedia.org/wiki/Contingency_table)) between them and the class label, and apply the Chi-squared test to them. [pandas.crosstab](http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.crosstab.html) can create contingency tables, and [scipy.stats.chi2_contingency](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chi2_contingency.html) can calculate the Chi-squared statistic for them.

There are 9 categorical features - as with the t-test, try to find one where the Chi-squared test returns an extreme result (rejecting the null that the data are independent), and one where it is less extreme.

**NOTE** - "less extreme" just means smaller test statistic/larger p-value. Even the least extreme differences may be strongly statistically significant.

Your *main* goal is the hypothesis tests, so don't spend too much time on the exploration/visualization piece. That is just a means to an end - use simple visualizations, such as boxplots or a scatter matrix (both built in to pandas), to get a feel for the overall distribution of the variables.

This is challenging, so manage your time and aim for a baseline of at least running two t-tests and two Chi-squared tests before polishing. And don't forget to answer the questions in part 3, even if your results in this part aren't what you want them to be.

In [62]:
# TODO

# Compute the continuous valued feature list
feat_cont = floats

#Compute the categorical valued feature list
#Remove the class label column(A16)
feat_cat = list(set(col_names) - set(feat_cont))
feat_cat.remove('A16')

print(feat_cont, feat_cat)

['A2', 'A3', 'A8', 'A11', 'A14', 'A15'] ['A7', 'A10', 'A5', 'A12', 'A13', 'A1', 'A4', 'A6', 'A9']


In [46]:
#Group by the credit class attribute(A16)
grouped = credit_df.groupby(credit_df.A16)

#Extract the entries in the positive class
pos_credit_df = grouped.get_group('+').drop(columns=['A16'])

#Extract the entries in the negative class
neg_credit_df = grouped.get_group('-').drop(columns=['A16'])

# Verify the entry counts in the positive and negative class dataframes
# by computing the the '+' and '-' value counts.
display(credit_df.A16.value_counts())
display(pos_credit_df)
display(neg_credit_df)


-    383
+    307
Name: A16, dtype: int64

A1     A2     A3   A4   A5   A6   A7  ...  A9 A10   A11  A12 A13    A14     A15
0    b  30.83  0.000    u    g    w    v  ...   t   t   1.0    f   g  202.0     0.0
1    a  58.67  4.460    u    g    q    h  ...   t   t   6.0    f   g   43.0   560.0
2    a  24.50  0.500    u    g    q    h  ...   t   f   0.0    f   g  280.0   824.0
3    b  27.83  1.540    u    g    w    v  ...   t   t   5.0    t   g  100.0     3.0
4    b  20.17  5.625    u    g    w    v  ...   t   f   0.0    f   s  120.0     0.0
..  ..    ...    ...  ...  ...  ...  ...  ...  ..  ..   ...  ...  ..    ...     ...
599  b  20.50  2.415    u    g    c    v  ...   t   t  11.0    t   g  200.0  3000.0
600  a  29.50  0.460    u    g    k    v  ...   t   t   4.0    f   g  380.0   500.0
606  b  16.17  0.040    u    g    c    v  ...   f   f   0.0    f   g    0.0     0.0
621  b  22.67  0.165    u    g    c    j  ...   f   f   0.0    t   s    0.0     0.0
622  a  25.58  0.000  NaN  NaN  NaN  NaN  ...   f   f   0.0    f   p    NaN     0.0

[307 rows x 15 columns]

A1     A2      A3 A4 A5  A6  A7      A8 A9 A10  A11 A12 A13    A14    A15
70   b  32.33   7.500  u  g   e  bb   1.585  t   f  0.0   t   s  420.0    0.0
71   b  34.83   4.000  u  g   d  bb  12.500  t   f  0.0   t   g    NaN    0.0
72   a  38.58   5.000  u  g  cc   v  13.500  t   f  0.0   t   g  980.0    0.0
73   b  44.25   0.500  u  g   m   v  10.750  t   f  0.0   f   s  400.0    0.0
74   b  44.83   7.000  y  p   c   v   1.625  f   f  0.0   f   g  160.0    2.0
..  ..    ...     ... .. ..  ..  ..     ... ..  ..  ...  ..  ..    ...    ...
685  b  21.08  10.085  y  p   e   h   1.250  f   f  0.0   f   g  260.0    0.0
686  a  22.67   0.750  u  g   c   v   2.000  f   t  2.0   t   g  200.0  394.0
687  a  25.25  13.500  y  p  ff  ff   2.000  f   t  1.0   t   g  200.0    1.0
688  b  17.92   0.205  u  g  aa   v   0.040  f   f  0.0   f   g  280.0  750.0
689  b  35.00   3.375  u  g   c   h   8.290  f   f  0.0   t   g    0.0    0.0

[383 rows x 15 columns]

In [54]:
from scipy.stats import ttest_ind

pvalues = []

# Use 2 sample ttest to see if there is a significant difference for a feature 
# between the 2 credit classes. 
# Store the (feature name, pvalue) tuple in the pvalues list
for feat in feat_cont:
  pvalues.append((feat,ttest_ind(pos_credit_df[feat], neg_credit_df[feat], nan_policy='omit').pvalue))

#Sort the pvalues list on the pvalue for each feature
pvalues.sort(key=lambda x:x[1])
display(pvalues)

[('A11', 7.957718568079967e-29),
 ('A8', 3.6710537401601785e-18),
 ('A3', 4.551680702308068e-08),
 ('A15', 3.4520256956287944e-06),
 ('A2', 2.0276637071781407e-05),
 ('A14', 0.008586135473979569)]

In [64]:
from scipy.stats import chi2_contingency

chi2_pvalues = []

# Use the chi2_contingency function to see is there is any correlation
# between a feature and the credit class
# Store the (feature name, pvalue) tuple in the chi2_pvalues list
for feat in feat_cat:
  _,pvalue,_,_ = chi2_contingency(pd.crosstab(credit_df[feat], credit_df.A16))
  chi2_pvalues.append((feat, pvalue))

#Sort the chi2_pvalues list on the pvalue for each feature
chi2_pvalues.sort(key=lambda x:x[1])
display(chi2_pvalues)

[('A9', 3.1185900878457007e-79),
 ('A10', 5.675727374527571e-33),
 ('A6', 3.4999300402715717e-15),
 ('A7', 3.62545287237226e-07),
 ('A5', 2.010680204180363e-06),
 ('A4', 2.010680204180363e-06),
 ('A13', 0.010094291370456362),
 ('A12', 0.45094587758631943),
 ('A1', 0.5768937883001117)]

## Part 3 - Analysis and Interpretation

Now that you've looked at the data, answer the following questions:

- Interpret and explain the two t-tests you ran - what do they tell you about the relationships between the continuous features you selected and the class labels?
- Interpret and explain the two Chi-squared tests you ran - what do they tell you about the relationships between the categorical features you selected and the class labels?
- What was the most challenging part of this sprint challenge?

Answer with text, but feel free to intersperse example code/results or refer to it from earlier.

Analysis of t-test results:

We assume a confidence level of 99%, which implies a cutoff pvalue of 
0.01(1-0.99)

All the pvalues are less than 0.01 and therefore the Null hypothesis can be rejected for all of them i.e. all that attributes have significant statistical differences between the positive and negative credit classes.


Feature A11, with the lowest pvalue, is extremely different between the 2 classes.

Feature A14,  with the highest pvalue, is notably less different(as compared to A11) even though it still has a significant statistical difference between the 2 classes.

Analysis of chi2 test results:

We, again, assume a confidence level of 99%, which implies a cutoff pvalue of 0.01(1-0.99).

Therefore we cannot reject the Null hypothesis that there is no relationship between the class label and features for A13(The pvalue here barely makes the cutoff and shouldn't really be considered part of the "unable to reject" feature list?), A12 and A1.

Feature A9, with the lowest pvalue, is the most extreme case where we can reject the Null hypothesis that there is no relationship between it and the class label.
Feature A4, is a less extreme case where we  can still reject the Null hypothesis.

## Part 4 - Bayesian vs Frequentist Statistics

Using a minimum of 2-3 sentences, give an explanation of Bayesian and Frequentist statistics, and then compare and contrast these two approaches to statistical inference.



Frequentist statistics computes probabilities using observable counts/frequencies e.g. there are six numbers on a die, so the probability of getting
any one of those numbers(assuming a fair die) should be 1/6; throw a die N times and observe the count of the number of times each of the numbers appear $n_1$,...,$n_6$, the probability $p_i$ for each number will be $p_i$ = $n_i$/N.

Bayesian statistics tell you how to compute the probability of an event(H), given that another event(E) has occurred by updating the prior(previously known) probability of H to incorporate the fact that E has occurred.

Bayesian and Frequentist statistics are complementary in the sense that the prior probabilities used in Bayesian statistics are normally provided by Frequentist techniques.

# Stretch Goals: 
Do these to get a 3. These are not required in order to pass the Sprint Challenge.

## Part 1: 

Make sure that all of your dataframe columns have the appropriate data types. *Hint:* If a column has the datatype of "object" even though it's made up of float or integer values, you can coerce it to act as a numeric column by using the `pd.to_numeric()` function. In order to get a 3 on this section make sure that your data exploration is particularly well commented, easy to follow, and thorough.

## Part 2:

Write functions that can calculate t-tests and chi^2 tests on all of the appropriate column combinations from the dataset. (Remember that certain tests require certain variable types.)

## Part 3: 

Calculate and report confidence intervals on your most important mean estimates (choose at least two). Make some kind of a graphic or visualization to help us see visually how precise these estimates are.

## Part 4:

Give an extra awesome explanation of Bayesian vs Frequentist Statistics. Maybe use code or visualizations, or any other means necessary to show an above average grasp of these high level concepts.

In [ ]:
# You can work the stretch goals down here or back up in their regular sections
# just make sure that they are labeled so that we can easily differentiate
# your main work from the stretch goals.